File: WindowsFormsApplicationlaba2/WindowsFormsApplicationlaba2/Spiderswolf.cs
﻿using System;
using System.Collections.Generic;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace WindowsFormsApplicationlaba2
{
    public class Spiderswolf : Arthropods, IComparable<Spiderswolf>, IEquatable<Spiderswolf>
    {
        public override int MaxSpeed
        {
            get
            {
                return base.MaxSpeed;
            }

            protected set
            {
                if (value > 0 && value < 100)

                {
                    base.MaxSpeed = value;
                }
                else
                {
                    base.MaxSpeed = 30;
                }
            }
        }
        public override double Weight
        {
            get
            {
                return base.Weight;
            }
            protected set
            {
                if (value > 10 && value < 30)
                {
                    base.Weight = value;
                }
                else
                {
                    base.Weight = 20;
                }
            }
        }

        public Spiderswolf(int maxSpeed, double weight, Color color)
        {
            this.MaxSpeed = maxSpeed;
            this.ColorBody = color;
            this.Weight = weight;

            //Random rand = new Random();
            startPosX = 0;
            startPosY = 0;

        }
        public Spiderswolf(string info)
        {
            string[] strs = info.Split(';');
            if (strs.Length == 3)
            {
                Weight = Convert.ToInt32(strs[0]);
                MaxSpeed = Convert.ToInt32(strs[1]);
                ColorBody = Color.FromName(strs[2]);
            }
            Random rand = new Random();
            startPosX = rand.Next(10, 200);
            startPosY = rand.Next(10, 200);
        }
        public override void drawSpiderwolf(Graphics g)
        {
            drawSpider(g);
        }



        protected virtual void drawSpider(Graphics g)
        {

            Brush brush = new SolidBrush(ColorBody);
            g.FillEllipse(brush, startPosX + 50, startPosY + 10, 30, 30);
            g.FillEllipse(brush, startPosX + 45, startPosY + 30, 40, 40);
            Pen pen = new Pen(ColorBody);
            g.DrawLine(pen, startPosX + 35, startPosY + 30, startPosX + 50, startPosY + 40);
            g.DrawLine(pen, startPosX + 30, startPosY + 45, startPosX + 50, startPosY + 50);
            g.DrawLine(pen, startPosX + 35, startPosY + 65, startPosX + 50, startPosY + 60);
            g.DrawLine(pen, startPosX + 25, startPosY + 40, startPosX + 35, startPosY + 30);
            g.DrawLine(pen, startPosX + 20, startPosY + 55, startPosX + 30, startPosY + 45);
            g.DrawLine(pen, startPosX + 25, startPosY + 80, startPosX + 35, startPosY + 65);
            g.DrawLine(pen, startPosX + 40, startPosY + 17, startPosX + 53, startPosY + 31);
            g.DrawLine(pen, startPosX + 50, startPosY + 7, startPosX + 40, startPosY + 17);
            g.DrawLine(pen, startPosX + 75, startPosY + 34, startPosX + 95, startPosY + 17);
            g.DrawLine(pen, startPosX + 80, startPosY + 40, startPosX + 95, startPosY + 30);
            g.DrawLine(pen, startPosX + 95, startPosY + 17, startPosX + 85, startPosY + 7);
            g.DrawLine(pen, startPosX + 95, startPosY + 30, startPosX + 105, startPosY + 42);
            g.DrawLine(pen, startPosX + 80, startPosY + 53, startPosX + 100, startPosY + 47);
            g.DrawLine(pen, startPosX + 100, startPosY + 47, startPosX + 110, startPosY + 55);
            g.DrawLine(pen, startPosX + 80, startPosY + 60, startPosX + 96, startPosY + 65);
            g.DrawLine(pen, startPosX + 96, startPosY + 65, startPosX + 103, startPosY + 80);

        }

        public override void moveSpiderwolf(Graphics g)
        {
            startPosX += MaxSpeed * 50 / (float)Weight;
            drawSpiderwolf(g);
        }
        public override string getInfo()
        {
            return Weight + ";" + MaxSpeed + ";" + ColorBody.Name;
        }
        public int CompareTo(Spiderswolf other)
        {
            if (other == null)
            {
                return 1;
            }
            if (Weight != other.Weight)
            {
                return Weight.CompareTo(other.Weight);
            }
            if (MaxSpeed != other.MaxSpeed)
            {
                return MaxSpeed.CompareTo(other.MaxSpeed);
            }

            if (ColorBody!= other.ColorBody)
            {
                return ColorBody.Name.CompareTo(other.ColorBody.Name);
            }
            return 0;
        }

        public bool Equals(Spiderswolf other)
        {
            if (other == null)
            {
                return false;
            }
            if (Weight != other.Weight)
            {
                return false;
            }
            if (MaxSpeed != other.MaxSpeed)
            {
                return false;
            }

            if (ColorBody != other.ColorBody)
            {
                return false;
            }
            return true;
        }
        public override bool Equals(object obj)
        {
            if (obj == null)
            {
                return false;

            }
            Spiderswolf SpiderswolfObj = obj as Spiderswolf;
            if (SpiderswolfObj == null)
            {
                return false;
            }
            else
            {
                return Equals(SpiderswolfObj);
            }
        }
        public override int GetHashCode()
        {
            return MaxSpeed.GetHashCode();
        }
    }
}

File: TPlaba2/src/Spiderswolf.java
import java.awt.Color;
import java.awt.Graphics;

public class Spiderswolf extends Arthropods {
	public Spiderswolf(int weight, int maxspeed, Color color) {
		this.Weight = weight;
		this.MaxSpeed = maxspeed;

		this.ColorBody = color;

		startPosX = 100;
		startPosY = 100;
	}

	protected void setWeight(int w) {
		if (w > 0 && w < 34)
			super.Weight = w;
		else
			super.Weight = 15;
	}

	public int getWeight() {
		return super.Weight;
	}

	protected void setMaxSpeed(int p) {
		if (p > 0 && p < 100)
			super.MaxSpeed = p;
		else
			super.MaxSpeed = 30;
	}

	public int getMaxSpeed() {
		return super.MaxSpeed;
	}

	public void drawSpiderwolf(Graphics g) {
		drawSpider(g);
	}

	protected void drawSpider(Graphics g) {
		g.setColor(ColorBody);
		g.fillOval(startPosX + 50, startPosY + 8,  30, 30);
		g.fillOval(startPosX + 45, startPosY + 30,  40,  40);

		g.drawLine(startPosX + 35, startPosY + 30, startPosX +  50,startPosY +  40);
		g.drawLine(startPosX + 30, startPosY + 45, startPosX +  50, startPosY + 50);
		g.drawLine(startPosX + 35, startPosY + 65, startPosX + 50, startPosY + 60);
		g.drawLine(startPosX + 25, startPosY + 40, startPosX +  35, startPosY + 30);
		g.drawLine(startPosX + 20, startPosY + 55, startPosX +  30,startPosY +  45);
		g.drawLine(startPosX + 25, startPosY + 80,startPosX + 35, startPosY + 65);
		g.drawLine(startPosX + 40, startPosY + 17, startPosX +  53, startPosY + 31);
		g.drawLine(startPosX + 50, startPosY + 7, startPosX + 40, startPosY + 17);
		g.drawLine(startPosX + 75, startPosY + 34, startPosX +  95,startPosY + 17);
		g.drawLine(startPosX + 80, startPosY + 40,startPosX +  95, startPosY + 30);
		g.drawLine(startPosX + 95, startPosY + 17, startPosX +  85, startPosY + 7);
		g.drawLine(startPosX + 95, startPosY + 30, startPosX +  105, startPosY + 42);
		g.drawLine(startPosX + 80, startPosY + 53, startPosX + 100,startPosY +  47);
		g.drawLine(startPosX + 100, startPosY + 47,startPosX + 110, startPosY + 55);
		g.drawLine(startPosX + 80, startPosY + 60,startPosX +  96, startPosY + 65);
		g.drawLine(startPosX + 96, startPosY + 65, startPosX + 103, startPosY + 80);
	}

	public void moveSpiderwolf(Graphics g) {
		startPosX += MaxSpeed * 50 / (int) Weight;
		drawSpiderwolf(g);
	}
}
